File: app/src/main/java/com/dev/ome/corentin/xkcdapp/MainPresenter.java
package com.dev.ome.corentin.xkcdapp;

import android.arch.lifecycle.LiveData;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;

import java.util.Random;

public class MainPresenter {
    private MainView view;
    private transient ComicsDao dao;
    private transient XkcdAPI xkcdAPI;

    private transient LiveData<PagedList<Comic>> comicsList;

    public void setView(MainView view) {
        this.view = view;
    }

    public void setDao(ComicsDao dao) {
        this.dao = dao;
        this.comicsList = new LivePagedListBuilder<>(this.dao.getAllComics(), 30)
                .build();
    }

    public void setXkcdAPI(XkcdAPI xkcdAPI) {
        this.xkcdAPI = xkcdAPI;
    }

    public void onDeleteAll() {
        this.dao.deleteAll();
    }

    public void onAdd() {
        Comic comic = xkcdAPI.getComic(new Random().nextInt(1000));
        this.view.log(comic.toString());
        this.dao.insert(comic);
    }

    public LiveData<PagedList<Comic>> getComicsList() {
        return comicsList;
    }

    public void clickItem(Comic comic){
        this.view.openOnClickItem(comic);
    }
}

File: app/src/main/java/com/dev/ome/corentin/xkcdapp/MainPresenter.kt
package com.dev.ome.corentin.xkcdapp

import android.arch.lifecycle.LiveData
import android.arch.paging.LivePagedListBuilder
import android.arch.paging.PagedList

import java.util.Random

class MainPresenter {
    private var view: MainView? = null

    private var dao: ComicsDao? = null

    private var xkcdAPI: XkcdAPI? = null


    var comicsList: LiveData<PagedList<Comic>>? = null
        private set

    fun setView(view: MainView) {
        this.view = view
    }

    fun setDao(dao: ComicsDao) {
        this.dao = dao
        this.comicsList = LivePagedListBuilder(this.dao!!.allComics, 30)
                .build()
    }

    fun setXkcdAPI(xkcdAPI: XkcdAPI) {
        this.xkcdAPI = xkcdAPI
    }

    fun onDeleteAll() {
        this.dao!!.deleteAll()
    }

    fun onAdd() {
        val comic = xkcdAPI!!.getComic(Random().nextInt(1000))
        this.view!!.log(comic!!.toString())
        this.dao!!.insert(comic)
    }

    fun clickItem(comic: Comic) {
        this.view!!.openOnClickItem(comic)
    }
}
